test/infrastructure/docker/exp/docker/nodepool.go

```diff
@@ -83,67 +83,84 @@ func NewNodePool(kClient client.Client, cluster *clusterv1.Cluster, mp *clusterv
 			}
-			return errors.Wrap(err, "failed to reconcile machine")
+			machineDeleted = true
+			continue
 		}
+		totalNumberOfMachines++
 	}
-
-	return np.refresh()
-}
-
-// DeleteExtraMachines will delete all of the machines outside of the machine pool / docker machine pool spec and update
-// the docker machine pool status.
-func (np *NodePool) DeleteExtraMachines(ctx context.Context) error {
-	outOfSpecMachineNames := map[string]interface{}{}
-	for _, outOfSpecMachine := range np.machinesNotMatchingInfrastructureSpec() {
-		externalMachine, err := docker.NewMachine(np.cluster.Name, outOfSpecMachine.Name(), np.dockerMachinePool.Spec.Template.CustomImage, np.labelFilters, np.logger)
-		if err != nil {
-			return errors.Wrapf(err, "failed to create helper for managing the externalMachine named %s", outOfSpecMachine.Name())
+	if machineDeleted {
+		if err := np.refresh(); err != nil {
+			return errors.Wrapf(err, "failed to refresh the node pool")
 		}
+	}
 
-		if err := externalMachine.Delete(ctx); err != nil {
-			return errors.Wrapf(err, "failed to delete machine %s", outOfSpecMachine.Name())
+	// Add new machines if missing.
+	machineAdded := false
+	matchingMachineCount := len(np.machinesMatchingInfrastructureSpec())
+	if matchingMachineCount < desiredReplicas {
+		for i := 0; i < desiredReplicas-matchingMachineCount; i++ {
+			if err := np.addMachine(ctx); err != nil {
+				return errors.Wrap(err, "failed to create a new docker machine")
+			}
+			machineAdded = true
 		}
-
-		outOfSpecMachineNames[outOfSpecMachine.Name()] = nil
 	}
-
-	var stats []*infrav1exp.DockerMachinePoolInstanceStatus
-	for _, machineStatus := range np.dockerMachinePool.Status.Instances {
-		if _, ok := outOfSpecMachineNames[machineStatus.InstanceName]; !ok {
-			stats = append(stats, machineStatus)
+	if machineAdded {
+		if err := np.refresh(); err != nil {
+			return errors.Wrapf(err, "failed to refresh the node pool")
 		}
 	}
 
-	for _, overprovisioned := range stats[*np.machinePool.Spec.Replicas:] {
-		externalMachine, err := docker.NewMachine(np.cluster.Name, overprovisioned.InstanceName, np.dockerMachinePool.Spec.Template.CustomImage, np.labelFilters, np.logger)
-		if err != nil {
-			return errors.Wrapf(err, "failed to create helper for managing the externalMachine named %s", overprovisioned.InstanceName)
+	// First remove instance status for machines no longer existing, then reconcile the existing machines.
+	// NOTE: the status is the only source of truth for understanding if the machine is already bootstrapped, ready etc.
+	// so we are preserving the existing status and using it as a bases for the next reconcile machine.
+	instances := make([]*infrav1exp.DockerMachinePoolInstanceStatus, 0, len(np.machines))
+	for i := range np.dockerMachinePool.Status.Instances {
+		instance := np.dockerMachinePool.Status.Instances[i]
+		found := false
+		for j := range np.machines {
+			if instance.InstanceName == np.machines[j].Name() {
+				found = true
```

Comment:
You could remove `found`, add `break` here move the `append` operation when the machine is found

Reply:
done!